vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -0,0 +1,132 @@
+import enum
+from enum import Enum
+from typing import Any, Dict, List, Optional
+
+import torch
+from torch.nn.parameter import Parameter
+
+from vllm.model_executor.layers.linear import (LinearMethodBase,
+                                               set_weight_attrs)
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+
+
+class FP8Config(QuantizationConfig):
+    """Config class for FP8."""
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "fp8"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [
+            torch.bfloat16, torch.half, torch.float8_e4m3fn, torch.float8_e5m2
+        ]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 90
+
+    @classmethod
+    def get_config_filenames(cls) -> List[str]:
+        return []
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "FP8Config":
+        return cls()
+
+    def get_linear_method(self) -> "Fp8LinearMethod":
+        return Fp8LinearMethod(self)
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class Fp8LinearState(Enum):
+
+    UNINITIALIZED = enum.auto()
+    READY = enum.auto()
+
+
+class Fp8LinearMethod(LinearMethodBase):
+    """Linear method for FP8.
+
+    Args:
+        quant_config: The quantization config.
+    """
+
+    def __init__(self, quant_config: FP8Config):
+        self.quant_config = quant_config
+
+    def create_weights(
+        self,
+        layer: torch.nn.Module,
+        input_size_per_partition: int,
+        output_size_per_partition: int,
+        input_size: int,
+        output_size: int,
+        params_dtype: torch.dtype,
+        **extra_weight_attrs,
+    ):
+        weight = Parameter(torch.empty(output_size_per_partition,
+                                       input_size_per_partition,
+                                       dtype=params_dtype),
+                           requires_grad=False)
+        set_weight_attrs(weight, {"input_dim": 1, "output_dim": 0})
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(weight, extra_weight_attrs)
+
+        scale = Parameter(
+            torch.empty(1, dtype=torch.float32),
+            requires_grad=False,
+        )
+        layer.register_parameter("scale", scale)
+        set_weight_attrs(scale, extra_weight_attrs)
+        layer.fp8_linear_state = Fp8LinearState.UNINITIALIZED
+
+    def apply_weights(self,
+                      layer: torch.nn.Module,
+                      x: torch.Tensor,
+                      bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        qinput, scale = per_tensor_quantize(x)
+
+        if layer.fp8_linear_state == Fp8LinearState.UNINITIALIZED:
+            qweight, weight_scale = per_tensor_quantize(layer.weight)
+            layer.weight.data = qweight.t()
```

Comment:
Will this dealloc the memory allocated for the original full precision weight Parameter? Unfortunately this will not help reduce peak memory required since we will be allocating full precision weights for the entire model before quantizing down in the forward pass - but that is fine for now.

Reply:
Right this doesn't help with memory footprint at all. Like you said we have to quantize weights during weight loading, but it's not straightforward to store the scaling factor at that time.

Reply:
I think the right approach here is to have model files with weights that are already fp8 quantized and have the scaling factors in them -- I implemented this here https://github.com/vllm-project/vllm/pull/3954/files#diff-74473eb619768ed055303d29d74c47b1753615e8323e69467506b2b42d3f9898R139. Even if only the fully connected layers are quantized, the model files and memory use gets *a lot* smaller this way :)

Reply:
Yes that's the easiest way as it is compatible with all existing quantization methods. Ideally I wish we could make both methods working on the other hand.

Reply:
On my first question though, once the quantization has been done at first run, will the original weight be deallocated and the memory usage decreased? I'm not sure the difference between `layer.weight.data = qweight.t()` and `layer.weight = qweight.t()`

Reply:
The current code is actually problematic, because the weight parameter has data type float16, but we assigned a float8 tensor, so `layer.weight = qweight.t()` will throw an error because it does a validation.